### PR TITLE
Mega cleanup

### PR DESCRIPTION
**Pull request by ChristianBreitkreutz**

### CODE REVIEW

File: t/README.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
...
+
 # run all test with prove
-    prove -l -j 9
+    prove -rl -j9
 
-# install needed packages
-    cpanm --installdeps --with-develop .
\ No newline at end of file
+## helper for development
+Run first all the tests and perlcritic on severity level 3 for t and lib
```
Comment (cgebken) [inline t/README.md]: level 1

File: lib/Test/Mockify/Matcher.pm [hunk 2]
```diff
@@ -68,8 +69,8 @@ Use the Matcher Number() to Check for the string '0' (perl cannot differ that)
 =cut
 sub String(;$) {
     my ($Value) = @_;
-    die('NotAString') if $Value && !IsString($Value);
-    die('Use the Matcher Number() to Check for the string \'0\' (perl cannot differ that)') if defined $Value && $Value eq '0';
+    Error('NotAString') if $Value && !IsString($Value);
+    Error("Please use the Matcher Number($Value) to Check for the string '$Value' (perl can not differ numbers and strings)") if defined $Value && IsFloat($Value);
```
Comment (cgebken) [inline lib/Test/Mockify/Matcher.pm]: "Please use the Matcher Number($Value) to check the string '$Value' (perl can not distinguish between numbers and strings)"

File: lib/Test/Mockify/Method.pm [hunk 3]
```diff
@@ -59,11 +60,11 @@ sub when {
     my @Signature;
     foreach my $Signature (keys %{$self->{'TypeStore'}}){
         if($Signature eq 'UsedWithWhenAny'){
-            die('It is not possible to use a mixture between "when" and "whenAny"');
+            Error('It is not possible to use a mixture between "when" and "whenAny"');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm]: It is not possible to mix "when" and "whenAny" for the same method.

File: lib/Test/Mockify/Method.pm [hunk 4]
```diff
@@ -81,18 +82,19 @@ It is not possible to mix C<whenAny> with C<when>.
 =cut
 sub whenAny {
     my $self = shift;
-    die ('"whenAny" don`t allow any parameters' ) if (@_);
+    Error ('"whenAny" don`t allow any parameters' ) if (@_);
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm]: "whenAny" doesn't allow any parameters

File: lib/Test/Mockify/Method.pm [hunk 5]
```diff
@@ -81,18 +82,19 @@ It is not possible to mix C<whenAny> with C<when>.
...
     if((scalar keys %{$self->{'TypeStore'}})){
-        die('"whenAny" can only used once. Also it is not possible to use a mixture between "when" and "whenAny"');
+        Error('"whenAny" can only used once. Also it is not possible to use a mixture between "when" and "whenAny"');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm]: You can use "whenAny" only once. Additionaly, it is not possible to mix "when" and "whenAny" for the same method.

File: lib/Test/Mockify/Method.pm [hunk 6]
```diff
@@ -104,7 +106,7 @@ sub _checkExpectedParameters{
 
     foreach my $ExistingParameter (@{$self->{'TypeStore'}{$SignatureKey}}){
         if($ExistingParameter->compareExpectedParameters($NewExpectedParameters)){
-            die('It is not possible two add two times the same method Signature.');
+            Error('It is not possible two add two times the same method Signature.');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm]: You can use a method signature only once.

File: lib/Test/Mockify/Method.pm [hunk 7]
```diff
@@ -115,11 +117,11 @@ sub _testMatcherStore {
     my ($MatcherStore, $NewExpectedParameterValue) = @_;
     if( defined $NewExpectedParameterValue->{'Value'} ){
         if($MatcherStore and not $MatcherStore->{'Value'}){
-            die('It is not possibel to mix "expected parameter" with previously set "any parameter".');
+            Error('It is not possibel to mix "expected parameter" with previously set "any parameter".');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm:120]: It is not possible to mix "expected parameter" and "any parameter".

File: lib/Test/Mockify/Method.pm [hunk 8]
```diff
@@ -115,11 +117,11 @@ sub _testMatcherStore {
...
         }
     } else {
         if($MatcherStore && $MatcherStore->{'Value'}){
-            die('It is not possibel to mix "any parameter" with previously set "expected parameter".');
+            Error('It is not possibel to mix "any parameter" with previously set "expected parameter".');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm:124]: It is not possible to mix "any parameter" and "expected parameter".

File: lib/Test/Mockify/Method.pm [hunk 9]
```diff
@@ -130,10 +132,10 @@ sub _testAnyStore {
     my ($AnyStore, $Type) = @_;
     if($AnyStore){
         if($AnyStore eq 'any' and $Type ne 'any'){
-            die('It is not possibel to mix "specific type" with previously set "any type".');
+            Error('It is not possibel to mix "specific type" with previously set "any type".');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm:135]: It is not possible to mix "specific type" and "any type".

File: lib/Test/Mockify/Method.pm [hunk 10]
```diff
@@ -130,10 +132,10 @@ sub _testAnyStore {
...
         }
         if($AnyStore ne 'any' and $Type eq 'any'){
-            die('It is not possibel to mix "any type" with previously set "specific type".');
+            Error('It is not possibel to mix "any type" with previously set "specific type".');
```
Comment (cgebken) [inline lib/Test/Mockify/Method.pm:138]: It is not possible to mix "any type" and "specific type".

File: lib/Test/Mockify/Tools.pm [hunk 11]
```diff
@@ -33,15 +33,16 @@ sub IsValid {
...
 
+    $hAdditionalInfo //= {};
     Error('Path or Object is needed') unless defined $PathOrObject;
     Error('Method name is needed') unless defined $MethodName;
     if( not $PathOrObject->can( $MethodName ) ){
         if( IsValid( ref( $PathOrObject ) ) ){
             $PathOrObject = ref( $PathOrObject );
         }
-        Error( $PathOrObject." donsn't have a method like: $MethodName", {'Method' => $MethodName});
+        Error( $PathOrObject." donsn't have a method like: $MethodName", {'Method' => $MethodName, %{$hAdditionalInfo}});
```
Comment (cgebken) [inline lib/Test/Mockify/Tools.pm]: The method $MethodName doesn't exist. (Maybe: Create a new method or choose another method.)

File: README.md [hunk 12]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
 
 ## mockStatic
 
-Sometimes it is not possible to inject the dependencies from the outside. This is especially the case when the package uses imports of static functions.
+Sometimes it is not possible to inject the dependencies from the outside. This is especially the case when the system-under-test uses functions with a fully qualified path.
```
Comment (cgebken) [inline README.md]: System Under Test

File: README.md [hunk 13]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
...
 `mockStatic` provides the possibility to mock static functions inside the mock/sut.
 
+Attention: The mocked function is valid as long as the $Mockify is defined. If You leave the scope or set the $Mockify to undef the injected method will be released.
```
Comment (cgebken) [inline README.md]: Attention: The mocked function is only valid if the $Mockify is defined. As soon as you leave the scope or set the $Mockify to `undef`, the injected method will be released.

File: README.md [hunk 14]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
...
+        shift;
+        if(Magic::Tools::Rabbit('black')){
+            return 1;
+        }else{
+            return 0;
+        }
+    }
+    1;
+
+In the Test it can be mocked like:
```
Comment (cgebken) [inline README.md]: In the test it can be mocked like this:

File: README.md [hunk 15]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
...
+        $Mockify->mockStatic('Magic::Tools::Rabbit')->when(String('black'))->thenReturn(1);
+        $Mockify->spy('log')->when(String());
+        my $SUT = $Mockify->getMockObject();
+
+        is($SUT->pullCylinder('black'), 1);
+        is(Magic::Tools::Rabbit('black'), 1); 
+    } # end scope
+    is(Magic::Tools::Rabbit('black'), 'someValue'); # The orignal method in in place again
+
+It can be mixed with normal `spy` and `mock`
```
Comment (cgebken) [inline README.md]: You can mix it with normal `spy` and `mock`.

File: README.md [hunk 16]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
...
+
+#### ACKNOWLEDGEMENTS
+Thanks to @dbucky for this amazing idea
+
+## mockImported
+
+Sometimes it is not possible to inject the dependencies from the outside. This is especially the case when the package uses imports of static functions.
+`mockImported` provides the possibility to mock imported functions inside the mock/sut.
```
Comment (cgebken) [inline README.md]: `mockImported` provides the possibility to mock these imported functions inside the mock/sut.

File: README.md [hunk 17]
```diff
@@ -68,14 +70,54 @@ For possible return types please look in [Test::Mockify::ReturnValue](https://me
...
+
+Unlike `mockStatic` is the injection with `mockImported` only in the mock valid.
```
Comment (cgebken) [inline README.md]: Unlike `mockStatic`, the injection with `mockImported` is only valid in the mock.

File: README.md [hunk 18]
```diff
@@ -86,18 +128,48 @@ Sometimes it is not possible to inject the dependencies from the outside. This i
 In the Test it can be mocked
```
Comment (cgebken) [inline README.md]: In the test it can be mocked like this:

File: README.md [hunk 19]
```diff
@@ -86,18 +128,48 @@ Sometimes it is not possible to inject the dependencies from the outside. This i
...
+    my $Mockify = Test::Mockify->new( 'SUT', [] );
+    $Mockify->mockImported('Magic::Tools','Rabbit')->when(String('white'))->thenReturn(1);
 
-    my $SUT = $MockObjectBuilder->getMockObject();
+    my $SUT = $Mockify->getMockObject();
     is($SUT->pullCylinder(), 1);
+    Rabbit('white');# return original result
     1;
 
 It can be mixed with normal `spy` and `mock`
```
Comment (cgebken) [inline README.md]: You can mix it with normal spy and mock.

File: README.md [hunk 20]
```diff
@@ -86,18 +128,48 @@ Sometimes it is not possible to inject the dependencies from the outside. This i
...
 
-#### Thx
-to @dbucky for this amazing idea
+## spyImported
+
+`spyImported` provides the possibility to spy imported functions inside the mock/sut.
+
+Unlike `spyStatic` is the injection with `spyImported` only in the mock valid.
```
Comment (cgebken) [inline README.md]: Unlike `spyStatic`, the injection with `spyImported` is only valid in the mock.

File: README.md [hunk 21]
```diff
@@ -86,18 +128,48 @@ Sometimes it is not possible to inject the dependencies from the outside. This i
...
+        shift;
+        if(Rabbit('white')){
+            return 1;
+        }else{
+            return 0;
+        }
+    }
+    1;
+
+In the Test it can be mocked
```
Comment (cgebken) [inline README.md]: In the test it can be mocked like this:

File: README.md [hunk 22]
```diff
@@ -86,18 +128,48 @@ Sometimes it is not possible to inject the dependencies from the outside. This i
...
+    use Magic::Tools qw ( Rabbit );
+    my $Mockify = Test::Mockify->new( 'SUT', [] );
+    $Mockify->spyImported('Magic::Tools','Rabbit')->when(String());
+
+    my $SUT = $Mockify->getMockObject();
+    is($SUT->pullCylinder(), 'SomeValue');
+    is(GetCallCount($SUT, 'Rabbit'), 1);
+    1;
+
+It can be mixed with normal `spy` and `mock`
```
Comment (cgebken) [inline README.md]: You can mix it with normal spy and mock.

File: README.md [hunk 23]
```diff
@@ -131,13 +203,13 @@ It is not possible to mix `whenAny` and `when` for the same method.
 
 ## spyStatic
 
-Provides the possibility to spy static functions inside the mock/sut.
+Provides the possibility to spy static functions around the mock/sut.
```
Comment (cgebken) [inline README.md]: Instead of around: connected to/ of ?

File: README.md [hunk 24]
```diff
@@ -148,62 +220,19 @@ Provides the possibility to spy static functions inside the mock/sut.
 In the Test it can be mocked
```
Comment (cgebken) [inline README.md]: In the test it can be mocked like this:

File: README.md [hunk 25]
```diff
@@ -148,62 +220,19 @@ Provides the possibility to spy static functions inside the mock/sut.
...
+    my $SUT = $Mockify->getMockObject();
 
     $SUT->pullCylinder();
-    is(GetCallCount($SUT, 'pullCylinder), 1);
+    Magic::Tools::Rabbit('black');
+    is(GetCallCount($SUT, 'Magic::Tools::Rabbit'), 2); # count as long as $Mockify is valid
 
     1;
 
 It can be mixed with normal `spy` and `mock`. For more options see, `mockStatic`
```
Comment (cgebken) [inline README.md]: You can mix it with normal spy and mock. For more options, see ...

File: t/ExampleProject/README.md [hunk 26]
```diff
@@ -0,0 +1,71 @@
+# Example project
+
+This is the circus example. On the stage are two shows. A magician and some clowns.
+
+## The !! MAGIC SHOW !!
+
+Summary: You can see a magician who pulls a rabbit out of a cylinder. But sometimes its a bit snappy.
```
Comment (cgebken) [inline t/ExampleProject/README.md]: it's

File: t/ExampleProject/README.md [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+
+### lib
+
+The `rabbit.pm` and the `Magician.pm` are both classes. So they have a constructor and methods with a `$self` context.
+
+### test
+
+* `t/test_MagicShow_Rabbit.t` Prove that the `rabbit.pm` behaves as expected.
```
Comment (cgebken) [inline t/ExampleProject/README.md]: proves

File: t/ExampleProject/README.md [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+* `t/test_MagicShow_Magician.t`
+  * The `rabbit.pm` is mocked in order to control the snappy behavior when testing the `Magician.pm`.
+  * This mock is injected into the `Magician.pm`.
+  * Unit test the behavior of the code in `Magician.pm`.
```
Comment (cgebken) [inline t/ExampleProject/README.md]: Unit tests are used to test the behavior of...

File: t/ExampleProject/README.md [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+
+# The !! KIDS SHOW !!
+
+Summary: They are two clowns, an old and a new one. Both are sitting on a seesaw and the heaviest will win.
```
Comment (cgebken) [inline t/ExampleProject/README.md]: On the stage are two clowns,

File: t/ExampleProject/README.md [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+The seesaw is a very modern one, but both of them remember the good old times when they used a timber beam for the kids show.
+
+### lib
+
+The `OldClown.pm` is a perl package with a function. The function have to be called with a fully qualified path. Here `t::ExampleProject::KidsShow::OldClown::BeHeavy`
```
Comment (cgebken) [inline t/ExampleProject/README.md]: The function has to be called with a fully qualified path, in this case:

File: t/ExampleProject/README.md [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+The `NewClown.pm` is a perl package with a exported function. The function have to be called with the imported function name. Here `ShowOfWeight`
```
Comment (cgebken) [inline t/ExampleProject/README.md]: an exported; The function has to be...

File: README.md [hunk 32]
```diff
@@ -27,6 +27,8 @@ Test::Mockify - minimal mocking framework for perl
 Use [Test::Mockify](https://metacpan.org/pod/Test::Mockify) to create and configure mock objects. Use [Test::Mockify::Verify](https://metacpan.org/pod/Test::Mockify::Verify) to
 verify the interactions with your mocks.
 
+You can find a Example Project in [ExampleProject](https://github.com/ChristianBreitkreutz/Mockify/tree/master/t/ExampleProject)
```
Comment (cgebken) [inline README.md]: an example project

**Comment by dbucky** [comment]
I found a problem when I was trying to incorporate this into a client project. It relates to incorrect parameter matching and retrieval for an instance function that has been statically mocked or spied. I came up with a test to illustrate the issue. I could use some help solving it, if you have time. https://github.com/dbucky/Mockify/commit/79711dd45557c84a014c7c0e32c753cb0f37d4b1

File: t/ExampleProject/README.md (hunk 31 above)
Comment (ChristianBreitkreutz) [reply t/ExampleProject/README.md]: ; or .?

File: lib/Test/Mockify/Method.pm (hunk 6 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Method.pm]: ok

File: lib/Test/Mockify/Method.pm (hunk 7 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Method.pm:120]: will not change

File: lib/Test/Mockify/Method.pm (hunk 8 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Method.pm:124]: will not change

File: lib/Test/Mockify/Method.pm (hunk 9 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Method.pm:135]: will not change

File: lib/Test/Mockify/Method.pm (hunk 10 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Method.pm:138]: will not change (but the typo ;-))

File: lib/Test/Mockify/Tools.pm (hunk 11 above)
Comment (ChristianBreitkreutz) [reply lib/Test/Mockify/Tools.pm]: fixed typo

File: t/ExampleProject/README.md (hunk 26 above)
Comment (ChristianBreitkreutz) [reply t/ExampleProject/README.md]: ok

File: t/ExampleProject/README.md (hunk 27 above)
Comment (ChristianBreitkreutz) [reply t/ExampleProject/README.md]: ok

**Review by ChristianBreitkreutz — COMMENTED** [review]
I forget something

**Pull request merged**